Mathlib/Analysis/Calculus/ImplicitContDiff.lean

```diff
@@ -0,0 +1,162 @@
+    apply IsCompl.of_eq
+    · ext x
+      rw [Submodule.mem_inf, Submodule.mem_bot, LinearMap.mem_ker, ContinuousLinearMap.coe_fst',
+        LinearMap.mem_ker]
+      constructor
+      · intro ⟨h1, h2⟩
+        rw [← Prod.mk.eta (p := x), h1] at h2
+        rw [Prod.ext_iff]
+        refine ⟨h1, h.bijective.injective ?_⟩
+        simp [h2, map_zero]
+      · rintro rfl
+        exact ⟨rfl, map_zero _⟩
+    · ext x
+      simp only [Submodule.mem_sup, Submodule.mem_top, iff_true]
+      obtain ⟨y, hy⟩ := h.bijective.surjective (f' x)
+      exact ⟨(0, y), by simp, x - (0, y), by simp [map_sub, ← hy], by abel⟩
+
+@[simp]
+lemma implicitFunctionData_leftFun_pt (h : IsContDiffImplicitAt n f f' a) :
+    h.implicitFunctionData.leftFun h.implicitFunctionData.pt = a.1 := rfl
+
+@[simp]
+lemma implicitFunctionData_rightFun_pt (h : IsContDiffImplicitAt n f f' a) :
+    h.implicitFunctionData.rightFun h.implicitFunctionData.pt = f a := rfl
+
+/-- The implicit function provided by the general theorem, from which we construct the more useful
+form `IsContDiffImplicitAt.implicitFunction`. -/
+noncomputable def implicitFunctionAux (h : IsContDiffImplicitAt n f f' a) : E → G → E × F :=
+  h.implicitFunctionData.implicitFunction
+
+lemma implicitFunctionAux_fst (h : IsContDiffImplicitAt n f f' a) :
+    ∀ᶠ p in 𝓝 (a.1, f a), (h.implicitFunctionAux p.1 p.2).1 = p.1 :=
+  h.implicitFunctionData.prod_map_implicitFunction.mono fun _ ↦ congr_arg Prod.fst
+
+lemma comp_implicitFunctionAux_eq_snd (h : IsContDiffImplicitAt n f f' a) :
+    ∀ᶠ p in 𝓝 (a.1, f a), f (h.implicitFunctionAux p.1 p.2) = p.2 :=
+  h.implicitFunctionData.prod_map_implicitFunction.mono fun _ ↦ congr_arg Prod.snd
+
+/-- Implicit function `φ` defined by `f (x, φ x) = f a`. -/
+noncomputable def implicitFunction (h : IsContDiffImplicitAt n f f' a) : E → F :=
+  fun x ↦ (h.implicitFunctionAux x (f a)).2
+
+lemma implicitFunction_def (h : IsContDiffImplicitAt n f f' a) :
+    h.implicitFunction = fun x ↦ (h.implicitFunctionData.implicitFunction.uncurry (x, (f a))).2 :=
+  rfl
+
+/-- `implicitFunction` is indeed the (local) implicit function defined by `f`. -/
+lemma apply_implicitFunction (h : IsContDiffImplicitAt n f f' a) :
+    ∀ᶠ x in 𝓝 a.1, f (x, h.implicitFunction x) = f a := by
+  have := h.comp_implicitFunctionAux_eq_snd
+  have hfst := h.implicitFunctionAux_fst
+  rw [nhds_prod_eq, eventually_swap_iff] at this hfst
+  apply this.curry.self_of_nhds.mp
+  apply hfst.curry.self_of_nhds.mono
+  simp_rw [Prod.swap_prod_mk]
+  intro x h1 h2
+  rw [← h2]
+  congr 1
+  ext
+  · rw [h1]
+  · rfl
+
+/-- If the implicit equation `f` is $C^n$ at `(x, y)`, then its implicit function `φ` around `x` is
+also $C^n$ at `x`. -/
+theorem contDiffAt_implicitFunction (h : IsContDiffImplicitAt n f f' a) :
+    ContDiffAt 𝕜 n h.implicitFunction a.1 := by
+  have := h.implicitFunctionData.contDiff_implicitFunction contDiffAt_fst h.contDiffAt h.one_le
+  rw [implicitFunction_def]
+  fun_prop
+
```

Comment:
Is it also easy from the implicit theorem API to get that, for all `y` close enough to `x`, if `f y = f a`, then `y` is of the form `(x, h.implicitFunction x)` (for some `x`  that you can even write down using the API?)
If it's easy, I think it's worth including.
If not, let's keep that for another PR.

Reply:
I'm sorry. Do you mean the statement:
> For all (x, y) close to (a, b), if f (x, y) = f (a, b), then y = φ x.

In other words, φ is locally unique.

I think the uniqueness theorem is missing from the more general version, so I will add that first, and then derive your statement from it. It feels more like a separate PR to me, especially when the current PR is already somewhat well refined (thanks to you and Michael).